cmd/test_load.go

```diff
@@ -0,0 +1,199 @@
+		logger.Debug("Detecting test type for...")
+		testType = detectTestType(lt.source.Data)
+	}
+
+	switch testType {
+	case testTypeJS:
+		logger.Debug("Trying to load as a JS test...")
+		runner, err := js.New(
+			gs.logger, lt.source, lt.fileSystems, lt.runtimeOptions, lt.builtInMetrics, lt.metricsRegistry,
+		)
+		// TODO: should we use common.UnwrapGojaInterruptedError() here?
+		if err != nil {
+			return fmt.Errorf("could not load JS test '%s': %w", testPath, err)
+		}
+		lt.initRunner = runner
+		return nil
+
+	case testTypeArchive:
+		logger.Debug("Trying to load test as an archive bundle...")
+
+		var arc *lib.Archive
+		arc, err := lib.ReadArchive(bytes.NewReader(lt.source.Data))
+		if err != nil {
+			return fmt.Errorf("could not load test archive bundle '%s': %w", testPath, err)
+		}
+		logger.Debugf("Loaded test as an archive bundle with type '%s'!", arc.Type)
+
+		switch arc.Type {
+		case testTypeJS:
+			logger.Debug("Evaluating JS from archive bundle...")
+			lt.initRunner, err = js.NewFromArchive(gs.logger, arc, lt.runtimeOptions, lt.builtInMetrics, lt.metricsRegistry)
+			if err != nil {
+				return fmt.Errorf("could not load JS from test archive bundle '%s': %w", testPath, err)
+			}
+			return nil
+		default:
+			return fmt.Errorf("archive '%s' has an unsupported test type '%s'", testPath, arc.Type)
+		}
+	default:
+		return fmt.Errorf("unknown or unspecified test type '%s' for '%s'", testType, testPath)
+	}
+}
+
+// readSource is a small wrapper around loader.ReadSource returning
+// result of the load and filesystems map
+func readSource(globalState *globalState, filename string) (*loader.SourceData, map[string]afero.Fs, error) {
+	pwd, err := globalState.getwd()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	filesystems := loader.CreateFilesystems(globalState.fs)
+	src, err := loader.ReadSource(globalState.logger, filename, pwd, filesystems, globalState.stdIn)
+	return src, filesystems, err
+}
+
+func detectTestType(data []byte) string {
+	if _, err := tar.NewReader(bytes.NewReader(data)).Next(); err == nil {
+		return testTypeArchive
+	}
+	return testTypeJS
+}
+
+func (lt *loadedTest) consolidateDeriveAndValidateConfig(
+	gs *globalState, cmd *cobra.Command,
+	cliConfGetter func(flags *pflag.FlagSet) (Config, error), // TODO: obviate
+) error {
+	var cliConfig Config
+	if cliConfGetter != nil {
+		gs.logger.Debug("Parsing CLI flags...")
+		var err error
+		cliConfig, err = cliConfGetter(cmd.Flags())
+		if err != nil {
+			return err
+		}
+	}
+
+	gs.logger.Debug("Consolidating config layers...")
+	consolidatedConfig, err := getConsolidatedConfig(gs, cliConfig, lt.initRunner.GetOptions())
+	if err != nil {
+		return err
+	}
+
+	gs.logger.Debug("Parsing thresholds and validating config...")
+	// Parse the thresholds, only if the --no-threshold flag is not set.
+	// If parsing the threshold expressions failed, consider it as an
+	// invalid configuration error.
+	if !lt.runtimeOptions.NoThresholds.Bool {
```

Comment:
What I like about this PR, is that it makes access the whole configuration, and especially `runtimeOptions` much easier without having to modify a bunch of APIs. Kudos for that 👍